p2p/dnsdisc/client.go

```diff
@@ -294,14 +295,56 @@ func (it *randomIterator) nextTree() *clientTree {
 	if len(it.trees) == 0 {
 		return nil
 	}
-	limit := rand.Intn(len(it.trees))
+
+	for {
+		// Find trees that might still have pending items to sync.
+		// If there are any, pick a random syncable tree.
+		syncable, disabled := it.syncableTrees()
+		if len(syncable) > 0 {
+			return syncable[rand.Intn(len(syncable))]
+		}
+		// The client tried all trees, and no sync action can be performed on any of them.
+		// The only meaningful thing to do now is waiting for any root record to get
+		// updated.
+		if !it.waitForRootUpdates(disabled) {
+			return nil // Iterator was closed.
+		}
+	}
+}
+
+// syncableTrees finds trees on which any meaningful sync action can be performed.
+func (it *randomIterator) syncableTrees() (syncable, disabled []*clientTree) {
```

Comment:
I still want optimize this to reuse the same slices over and over, but let's review the basic idea first.